src/darwin/Framework/CHIP/MTRBaseDevice.mm

```diff
@@ -2290,20 +2290,32 @@ + (MTRAttributeRequestPath *)requestPathWithEndpointID:(NSNumber * _Nullable)end
 
 - (BOOL)isEqualToAttributeRequestPath:(MTRAttributeRequestPath *)path
 {
-    if (!path)
+    if (!path) {
         return NO;
+    }
+
+    if (path == self) {
+        return YES;
+    }
+
+    // Note: Nil is a wildcard, and empty endpoint/cluster/attribute values imply nil as well.
+    // Pass iff we have objects that are pointer equivalent (eg: nil == nil || object_a == object_a), or they have values and the same logical values
+    if ((_endpoint == path.endpoint || (path.endpoint && [_endpoint isEqualToNumber:path.endpoint]))
+        && (_cluster == path.cluster || (path.cluster && [_cluster isEqualToNumber:path.cluster]))
+        && (_attribute == path.attribute || (path.attribute && [_attribute isEqualToNumber:path.attribute]))) {
+        return YES;
+    }
 
-    return (path.endpoint && [_endpoint isEqualToNumber:path.endpoint])
-        && (path.cluster && [_cluster isEqualToNumber:path.cluster])
-        && (path.attribute && [_attribute isEqualToNumber:path.attribute]);
+    return NO;
 }
 
 - (BOOL)isEqual:(id)object
 {
-    if ([object class] != [self class]) {
-        return NO;
+    if ([[object class] isKindOfClass:[MTRAttributeRequestPath class]]) {
```

Comment:
This is just wrong, it leads to incorrectly behaving isEqual when a sub-class is added (and if there are no sub-classes it behaves the same either way).

(Also, implementing isEqual correctly in a way that allows a sub-class instance with additional state to compare equal to a super class instance is way more complicated than just using `isKindOfClass:`. See e.g. http://www.angelikalanger.com/Articles/JavaSolutions/SecretsOfEquals/Equals.html and http://www.angelikalanger.com/Articles/JavaSolutions/SecretsOfEquals/Equals-2.html for a details discussion)

Reply:
This is not correct, but I'm going to remove this part of the PR, and we'll move on.